crates/astria-sequencer/src/app.rs

```diff
@@ -360,19 +374,50 @@ impl App {
     /// Since transaction execution now happens in the proposal phase, results
     /// are cached in the app and returned here during the usual ABCI block execution process.
     ///
+    /// If the proposal was not executed, the transaction will be executed.
+    ///
     /// Note that the first two "transactions" in the block, which are the proposer-generated
     /// commitments, are ignored.
-    #[instrument(name = "App::deliver_tx_after_execution", skip(self))]
-    pub(crate) fn deliver_tx_after_execution(
+    #[instrument(name = "App::deliver_tx_after_proposal", skip(self))]
+    pub(crate) async fn deliver_tx_after_proposal(
         &mut self,
-        tx_hash: &[u8; 32],
+        tx: abci::request::DeliverTx,
     ) -> Option<anyhow::Result<Vec<abci::Event>>> {
         if self.processed_txs < 2 {
             self.processed_txs += 1;
             return Some(Ok(vec![]));
         }
 
-        self.execution_result.remove(tx_hash)
+        if self.proposal_executed {
+            let tx_hash: [u8; 32] = sha2::Sha256::digest(&tx.tx).into();
+            return self.execution_result.remove(&tx_hash);
+        }
+
+        let Some(signed_tx) = raw::SignedTransaction::decode(&*tx.tx)
+            .map_err(|e| {
+                debug!(
+                    error = &e as &dyn std::error::Error,
+                    "failed to deserialize bytes as a signed transaction",
+                );
+                e
+            })
+            .ok()
+            .and_then(|raw_tx| {
+                SignedTransaction::try_from_raw(raw_tx)
+                    .map_err(|e| {
+                        debug!(
+                            error = &e as &dyn std::error::Error,
+                            "failed to convert raw signed transaction to native signed transaction"
+                        );
+                        e
+                    })
+                    .ok()
```

Comment:
this entire block looks a bit crazy. Could make a helper function to make this look cleaner:
```rust
fn signed_transaction_from_bytes(bytes: &[u8]) -> anyhow::Result<SignedTransaction> {
    let raw = raw::SignedTransaction::decode(bytes).context("failed to decode protobuf to signed transaction")?;
    let tx = SignedTransaction::try_from_raw(raw).context("failed to transform raw signed transaction to verified type")?;
    Ok(tx)
}
```
Then in the body:
```rust
let signed_tx = match signed_transaction_from_bytes(&tx.tx) {
    Err(e) => {
        debug!(
            error = AsRef::<dyn std::error::Error>::as_ref(&e),
            "failed to decode deliver tx payload to signed transaction; ignoring it",
        );
        return None;.
    }
    Ok(tx) => tx,
};
```

Reply:
this is copy-pasted from `execute_block_data` so we can just make a function used by both